Extend Termination Checking with Dummy Arguments

NathanielB123 opened the pull request:
Implements the termination checking extension suggested by @andreasabel in https://github.com/agda/agda/issues/7693#issuecomment-2621495501.

Really naive implementation that just adds a row/column of `=`/`?`s to every call matrix, as if every function had an extra dummy argument. This will obviously increases the size of all termination matrices, so probably has a performance impact (though it seems pretty small, and I suppose guardedness gets away with doing a very similar thing). 

The changes are pretty minimal, but feel quite hacky - also this is my first time attempting to contribute to Agda and I realise termination checking is a pretty critical component, so please check over carefully.

Most of the examples mentioned on that issue now typecheck, but a couple minimised ones such as
```agda
open import Agda.Builtin.Nat using (zero; suc; _+_) renaming (Nat to ℕ)

mutual
  f : (n : ℕ) → ℕ
  f (suc n) = g zero (f n)
  f _ = 42

  g : (m k : ℕ) → ℕ
  g m (suc k) = g m k
  g (suc zero) k = g zero (f 42)
  g _ _ = 42
```
still don't. I think this is reasonable though; the dummy argument in this case NEEDS to be a `ℕ` and specifically `zero` for termination to succeed (because `zero < suc zero`), unlike the other examples (where the dummy argument can be any base constructor). If Agda was to start considering all base constructors to be strictly smaller than non-base constructors, maybe turning the row of `=`s into a row of `=`/`<`s where appropriate would be sensible, and then cases like this should work; I have cooked up an even hackier implementation of exactly this at https://github.com/NathanielB123/agda/tree/base-lt.

NathanielB123 committed:
Extend TermCheck with dummy args

The fully-minimised test case still fails, but this might be
intended behaviour (because the dummy arg isn't necessarily
smaller than 'suc zero')

NathanielB123 committed:
Tweak test cases

Remove the minimised to 'f'/'g' one as I think it is reasonable to
fail there

NathanielB123 commented:
Ouch, looks like I broke something/this is not quite as easy as I hoped.

NathanielB123 committed:
Fix termination regression in std-lib

Turns out the new column of the call matrix should be '?'s not '='s

NathanielB123 commented:
Ok this is a little subtle - a column of `=`s doesn't work, but a column of `?`s is also incorrect because other base constructors should be considered equal to the dummy constructor (and I think this is actually necessary for the dummy argument to be of any real use).

NathanielB123 committed:
Build call matrix column properly

Also tweak tests again

NathanielB123 committed:
Move Issue1075.agda

NathanielB123 commented:
Looks like the only failing test now is https://github.com/agda/agda/blob/master/test/Fail/Issue1075.agda, which includes the comment "`-- I do not know whether it is supposed to terminate.`" I certainly don't know either, but I guess I'll move it to the success folder, lol

NathanielB123 committed:
Refactor into helper function

andreasabel reviewed (commented):
Thanks!  This looks reasonable.

Would be good to check that performance isn't affected in a significant way.

E.g. how does the termination checking time for `Issue1075` change?
(I suppose `--profile=internal` is the Agda option needed to display the accounting of runtime.)

Also, the comment on top of `Issue1075.agda` needs a new entry.  Currently it says:
```agda
-- 2014-03-06 Andreas, Reported by Fabien Renaud

-- This is a larger example for the termination checker, to test performance.
-- In 2014-02-X, it exhausted the heap.

-- 2013-03-17
-- The termination checker now rejects this code instead of crashing.
-- I do not know whether it is supposed to terminate.
```

NathanielB123 committed:
Add extra note to Issue1075

Also change date of previous note. The commit is from 2014-03-17
so I suspect this was a typo.

NathanielB123 commented:
> E.g. how does the termination checking time for `Issue1075` change? (I suppose `--profile=internal` is the Agda option needed to display the accounting of runtime.)

On my laptop (R5 5600H), typechecking `Issue1075.agda` on this branch seems to have gotten slightly _faster_

(This branch)
```
Checking Issue1075 (/home/nathaniel/agda/fork/agda/test/Succeed/Issue1075.agda).
Total                        1,200ms        
Miscellaneous                   28ms        
Coverage                       118ms (435ms)
Coverage.UnifyIndices          317ms        
Typing                          46ms (377ms)
Typing.CheckLHS                 75ms (175ms)
Typing.CheckLHS.UnifyIndices   100ms        
Typing.CheckRHS                 99ms        
Typing.TypeSig                  41ms        
Termination                      8ms (123ms)
Termination.Graph              107ms        
Parsing                         35ms  (96ms)
Parsing.OperatorsExpr           49ms        
Parsing.OperatorsPattern        11ms        
Deserialization                 65ms  (73ms)
Scoping                         13ms  (35ms)
Scoping.InverseScopeLookup      21ms        
Highlighting                    28ms        
Positivity                      18ms        
InterfaceInstantiateFull        11ms
```
vs

(Master)
```
Checking Issue1075 (/home/nathaniel/agda/fork/agda/test/Succeed/Issue1075.agda).
Total                        1,278ms        
Miscellaneous                   30ms        
Coverage                       121ms (456ms)
Coverage.UnifyIndices          335ms        
Typing                          58ms (395ms)
Typing.CheckLHS                 77ms (181ms)
Typing.CheckLHS.UnifyIndices   103ms        
Typing.CheckRHS                102ms        
Typing.TypeSig                  39ms        
Termination                     18ms (178ms)
Termination.Graph              144ms        
Termination.Compare             14ms        
Parsing                         36ms  (96ms)
Parsing.OperatorsExpr           49ms        
Parsing.OperatorsPattern        11ms        
Deserialization                 65ms  (73ms)
Scoping                         14ms  (26ms)
Scoping.InverseScopeLookup      12ms        
Highlighting                    24ms        
Positivity                      19ms        
/home/nathaniel/agda/fork/agda/test/Succeed/Issue1075.agda:287.1-439.54: error: [TerminationIssue]
Termination checking failed for the following functions:
...
```
(both built with `--optimise-heavily`)

but given the test succeeds on the new branch but fails previously, I'm not sure it's entirely fair. Compiling the standard library (`Everything.agda`) gives:

(This branch)
```
Total                                   449,879ms            
Miscellaneous                             2,711ms            
Typing                                   26,711ms (173,670ms)
Typing.CheckRHS                          54,052ms            
Typing.CheckLHS                          13,529ms  (36,293ms)
Typing.CheckLHS.UnifyIndices             22,763ms            
Typing.Generalize                        21,885ms            
Typing.OccursCheck                       14,189ms            
Typing.TypeSig                           13,966ms            
Typing.With                               5,667ms            
Typing.InstanceSearch                       140ms     (903ms)
Typing.InstanceSearch.FilterCandidates      491ms            
Typing.InstanceSearch.InitialCandidates     271ms            
Serialization                            54,175ms  (69,440ms)
Serialization.BinaryEncode                6,812ms            
Serialization.Compress                    5,970ms            
Serialization.Sort                        1,986ms            
Serialization.BuildInterface                495ms            
Coverage                                 34,240ms  (57,910ms)
Coverage.UnifyIndices                    23,670ms            
Scoping                                   5,033ms  (32,186ms)
Scoping.InverseScopeLookup               27,153ms            
Parsing                                   7,001ms  (23,986ms)
Parsing.OperatorsExpr                    14,139ms            
Parsing.OperatorsPattern                  2,845ms            
Deserialization                          15,155ms  (22,645ms)
Deserialization.Compaction                7,490ms            
Positivity                               21,027ms            
InterfaceInstantiateFull                 18,699ms            
DeadCode                                    332ms  (15,336ms)
DeadCode.DeadCodeReachable               15,003ms            
Termination                               1,592ms   (6,003ms)
Termination.RecCheck                      3,372ms            
Termination.Graph                           582ms            
Termination.Compare                         455ms            
Highlighting                              3,861ms            
Import                                    1,239ms            
Injectivity                                 867ms            
ProjectionLikeness                          242ms            
ModuleName                                   48ms
```
vs

(Master)
```
Total                                   414,393ms            
Miscellaneous                             2,612ms            
Typing                                   25,084ms (161,308ms)
Typing.CheckRHS                          50,574ms            
Typing.CheckLHS                          12,420ms  (33,858ms)
Typing.CheckLHS.UnifyIndices             21,437ms            
Typing.Generalize                        19,918ms            
Typing.OccursCheck                       13,026ms            
Typing.TypeSig                           12,905ms            
Typing.With                               5,098ms            
Typing.InstanceSearch                       132ms     (843ms)
Typing.InstanceSearch.FilterCandidates      464ms            
Typing.InstanceSearch.InitialCandidates     245ms            
Serialization                            50,331ms  (65,086ms)
Serialization.BinaryEncode                6,319ms            
Serialization.Compress                    5,720ms            
Serialization.Sort                        2,231ms            
Serialization.BuildInterface                483ms            
Coverage                                 31,879ms  (53,717ms)
Coverage.UnifyIndices                    21,838ms            
Scoping                                   4,794ms  (27,414ms)
Scoping.InverseScopeLookup               22,619ms            
Parsing                                   6,557ms  (22,299ms)
Parsing.OperatorsExpr                    13,131ms            
Parsing.OperatorsPattern                  2,610ms            
Deserialization                          14,216ms  (21,184ms)
Deserialization.Compaction                6,967ms            
Positivity                               19,822ms            
InterfaceInstantiateFull                 15,905ms            
DeadCode                                    319ms  (13,888ms)
DeadCode.DeadCodeReachable               13,568ms            
Termination                               1,613ms   (5,381ms)
Termination.RecCheck                      3,110ms            
Termination.Compare                         419ms            
Termination.Graph                           237ms            
Highlighting                              3,591ms            
Import                                    1,133ms            
Injectivity                                 780ms            
ProjectionLikeness                          224ms            
ModuleName                                   41ms
```
So looks like it has gotten slower, but `Termination` specifically is very similar, so maybe this is just random noise?

> Also, the comment on top of `Issue1075.agda` needs a new entry.

Ah yep good point - added

andreasabel commented:
Thanks for the experiment!            

> Compiling the standard library (`Everything.agda`) gives:

This shows an 8% total increase, which feels unlikely.  Could you try and run the experiment again?

NathanielB123 commented:
Annoyingly I'm getting pretty inconsistent results, perhaps due to on other apps running in the background. After closing everything but the terminal running Agda (through WSL, in case that is relevant) I got:

This branch:
```
Total                                   399,772ms            
Miscellaneous                             2,507ms            
Typing                                   24,202ms (154,827ms)
Typing.CheckRHS                          47,767ms            
Typing.CheckLHS                          11,936ms  (32,361ms)
Typing.CheckLHS.UnifyIndices             20,424ms            
Typing.Generalize                        19,342ms            
Typing.OccursCheck                       12,960ms            
Typing.TypeSig                           12,479ms            
Typing.With                               4,864ms            
Typing.InstanceSearch                       124ms     (849ms)
Typing.InstanceSearch.FilterCandidates      442ms            
Typing.InstanceSearch.InitialCandidates     282ms            
Serialization                            47,659ms  (61,627ms)
Serialization.BinaryEncode                6,393ms            
Serialization.Compress                    5,613ms            
Serialization.Sort                        1,500ms            
Serialization.BuildInterface                461ms            
Coverage                                 30,554ms  (51,763ms)
Coverage.UnifyIndices                    21,209ms            
Scoping                                   4,484ms  (26,604ms)
Scoping.InverseScopeLookup               22,120ms            
Parsing                                   6,023ms  (21,099ms)
Parsing.OperatorsExpr                    12,569ms            
Parsing.OperatorsPattern                  2,506ms            
Deserialization                          13,860ms  (20,556ms)
Deserialization.Compaction                6,695ms            
Positivity                               19,532ms            
InterfaceInstantiateFull                 17,004ms            
DeadCode                                    307ms  (13,085ms)
DeadCode.DeadCodeReachable               12,778ms            
Termination                               1,394ms   (5,601ms)
Termination.RecCheck                      3,238ms            
Termination.Graph                           551ms            
Termination.Compare                         417ms            
Highlighting                              3,412ms            
Import                                    1,123ms            
Injectivity                                 761ms            
ProjectionLikeness                          221ms            
ModuleName                                   40ms
```

Master branch:
```
Total                                   407,501ms            
Miscellaneous                             2,569ms            
Typing                                   25,365ms (159,105ms)
Typing.CheckRHS                          49,031ms            
Typing.CheckLHS                          12,390ms  (33,480ms)
Typing.CheckLHS.UnifyIndices             21,090ms            
Typing.Generalize                        19,687ms            
Typing.OccursCheck                       13,108ms            
Typing.TypeSig                           12,515ms            
Typing.With                               5,097ms            
Typing.InstanceSearch                       128ms     (818ms)
Typing.InstanceSearch.FilterCandidates      453ms            
Typing.InstanceSearch.InitialCandidates     236ms            
Serialization                            49,003ms  (62,541ms)
Serialization.BinaryEncode                5,908ms            
Serialization.Compress                    5,648ms            
Serialization.Sort                        1,534ms            
Serialization.BuildInterface                447ms            
Coverage                                 31,147ms  (52,602ms)
Coverage.UnifyIndices                    21,454ms            
Scoping                                   4,692ms  (26,612ms)
Scoping.InverseScopeLookup               21,920ms            
Parsing                                   6,208ms  (21,796ms)
Parsing.OperatorsExpr                    12,997ms            
Parsing.OperatorsPattern                  2,590ms            
Deserialization                          13,937ms  (20,690ms)
Deserialization.Compaction                6,753ms            
Positivity                               20,305ms            
InterfaceInstantiateFull                 16,912ms            
DeadCode                                    327ms  (13,479ms)
DeadCode.DeadCodeReachable               13,151ms            
Termination                               1,523ms   (5,177ms)
Termination.RecCheck                      3,012ms            
Termination.Compare                         409ms            
Termination.Graph                           232ms            
Highlighting                              3,501ms            
Import                                    1,169ms            
Injectivity                                 772ms            
ProjectionLikeness                          221ms            
ModuleName                                   43ms
```

Not sure this enough data for firm conclusions, but it does appear that, on this branch, when typechecking the standard library, `Termination.Graph` pretty reliably takes longer (from 200-250ms to 550-600ms), overall termination-checking time also increases (but by less proportionally), and overall typechecking time is basically unaffected (because termination checking is such a tiny fraction). I guess the decision of whether the trade-off makes sense is up to you (idk whether there are any scenarios in which `Termination`/`Termination.Graph` actually start to dominate runtime). I'm also happy to keep repeating the benchmark if you want further confirmation.

NathanielB123 committed:
Add another test case

NathanielB123 commented:
Closing due to problems expressed in https://github.com/agda/agda/issues/7693#issuecomment-2684625482 (i.e. this PR does fix the issue, but the solution is pretty hacky, and the existing the work-around isn't that painful)
